drivers/serial/uart_rpi_pico.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2021, Yonatan Schachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/uart.h>
+#include <drivers/pinctrl.h>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
+#include <hardware/uart.h>
+#include <hardware/gpio.h>
+#pragma GCC diagnostic pop
+
+struct uart_rpi_config {
+	uart_inst_t *const uart_dev;
+	uint32_t baudrate;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+static int uart_rpi_poll_in(const struct device *dev, unsigned char *c)
+{
+	const struct uart_rpi_config *config = dev->config;
+
+	if (!uart_is_readable(config->uart_dev)) {
+		return -1;
+	}
+
+	*c = (unsigned char)uart_get_hw(config->uart_dev)->dr;
+	return 0;
+}
+
+static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
+{
+	const struct uart_rpi_config *config = dev->config;
+
+	uart_putc_raw(config->uart_dev, c);
+}
+
+static int uart_rpi_init(const struct device *dev)
+{
+	const struct uart_rpi_config *config = dev->config;
+	int ret;
+
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return !uart_init(config->uart_dev, config->baudrate);
```

Comment:
this function returns:
```
\return Actual set baudrate
```
error handling has to be improved.

Reply:
Judging by the documentation and the implementation, it doesn't seem like this function can fail. So it should probably always return 0, but I'm not sure how error handling can be improved beyond that

Reply:
maybe (judging from docs)
```
if (uart_init(config->uart_dev, config->baudrate) != config->baudrate) {
    return -EIO
}
```

Reply:
The "actual" baudrate differs from the given baudrate because the peripheral can't produce any arbitrary clock rate, and has to rely on an approximation. For example, setting the baudrate to 115200 gives an actual baudrate of 115207.